### PR TITLE
refactor: use lua api for user commands and autocommands

### PR DESCRIPTION
**Pull request by sarmong**
I am still a bit new to autocommands, so please double-check that everything works

### CODE REVIEW

**Comment by alex-courtis** [comment]
We should update the minimum version of neovim to 0.7

lua API for :au and :map was not reliable in 0.6

**Review by alex-courtis — CHANGES_REQUESTED** [review]
~We should update the minimum version of neovim to 0.7~

~lua API for :au and :map was not reliable in 0.6~

Edit: I see that you have already done this!

**Review by alex-courtis — APPROVED** [review]

**Review by alex-courtis — APPROVED** [review]
Great update, thank you

**Comment by alex-courtis** [comment]
We should do a neovim version check when nvim-tree starts.

< 0.7 will result in a warning and nvim-tree does not start.

@kyazdani42 ?

**Comment by kyazdani42** [comment]
I'd prefer leaving a month to let people update before forcing this on everybody :) Some package managers might not have updated their neovim distribution yet.

**Comment by sarmong** [comment]
> I'd prefer leaving a month to let people update before forcing this on everybody :) Some package managers might not have updated their neovim distribution yet.

Treesitter updated a couple days after the 0.7 release. And it is very likely that 90% of nvim-tree users use treesitter as well. So they either already updated everything, or know how hold updates if they don't want to

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Tested most changes with 0.7.

Confirmed that it fails with 0.6.

We must add the startup version check before merging this.

File: lua/nvim-tree.lua [hunk 1]
```diff
@@ -272,40 +285,53 @@ function M.change_dir(name)
...
   end
   if opts.hijack_cursor then
-    vim.cmd "au CursorMoved NvimTree_* lua require'nvim-tree'.place_cursor_on_node()"
+    create_nvim_tree_autocmd("CursorMoved", { pattern = "NvimTree_*", callback = M.place_cursor_on_node })
   end
   if opts.update_cwd then
-    vim.cmd "au DirChanged * lua require'nvim-tree'.change_dir(vim.loop.cwd())"
+    create_nvim_tree_autocmd("DirChanged", {
+      callback = function()
+        M.change_dir(vim.loop.cwd)
```
Comment (kyazdani42) [inline lua/nvim-tree.lua:314]: missing function call to vim.loop.cwd

File: lua/nvim-tree/actions/open-file.lua [hunk 2]
```diff
@@ -256,13 +256,13 @@ function M.fn(mode, filename)
...
-          autocmd TextChangedI <buffer> setlocal bufhidden= | autocmd! RemoveBufHidden
-      augroup end
-    ]]
+
+      vim.api.nvim_create_autocmd({ "TextChanged", "TextChangedI" }, {
+        group = vim.api.nvim_create_augroup("RemoveBufHidden", {}),
+        buffer = vim.api.nvim_get_current_buf(),
+        command = "setlocal bufhidden=",
+        once = true,
+      })
```
Comment (kyazdani42) [inline lua/nvim-tree/actions/open-file.lua:265]: nice

**Review by kyazdani42 — CHANGES_REQUESTED** [review]

File: lua/nvim-tree/actions/open-file.lua (hunk 2 above)
Comment (gegoune) [reply lua/nvim-tree/actions/open-file.lua:265]: That command  can be replaced with callback and use api to set bufhidden I think. :)

File: lua/nvim-tree.lua (hunk 1 above)
Comment (sarmong) [reply lua/nvim-tree.lua]: updated

File: lua/nvim-tree/actions/open-file.lua (hunk 2 above)
Comment (sarmong) [reply lua/nvim-tree/actions/open-file.lua]: updated

File: lua/nvim-tree.lua [hunk 3]
```diff
@@ -469,6 +495,10 @@ local function validate_options(conf)
 end
 
 function M.setup(conf)
+  if vim.fn.has "nvim-0.7" == 1 then
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:498]: This should check for 0 return.

Comment (alex-courtis) [inline lua/nvim-tree.lua:498]: nvim greater than 0.7 is OK.

**Review by alex-courtis — CHANGES_REQUESTED** [review]

File: lua/nvim-tree.lua (hunk 3 above)
Comment (alex-courtis) [inline lua/nvim-tree.lua:498]: This test is backwards; 1 indicates success.

**Review by alex-courtis — CHANGES_REQUESTED** [review]

File: lua/nvim-tree.lua (hunk 3 above)
Comment (sarmong) [reply lua/nvim-tree.lua]: Right, thank you! Changed

File: lua/nvim-tree.lua [hunk 4]
```diff
@@ -469,6 +495,10 @@ local function validate_options(conf)
 end
 
 function M.setup(conf)
+  if vim.fn.has "nvim-0.7" == 0 then
+    utils.warn "nvim-tree.lua requires Neovim 0.7 or higher"
```
Comment (kyazdani42) [inline lua/nvim-tree.lua:499]: maybe return at this point ?

**Review by kyazdani42 — COMMENTED** [review]
almost good :) Could you rebase and manage the autocmd for coc diagnostic updates that has been just added ? 
Also could you replace all the `vim.api` to `api` ? It's declared on top of the file :)

File: lua/nvim-tree.lua (hunk 4 above)
Comment (sarmong) [reply lua/nvim-tree.lua:506]: done

**Review by kyazdani42 — APPROVED** [review]
LGTM, @alex-courtis can you make a last review ? Then we'll merge.
I think i'll add a new issue for breaking change.

File: lua/nvim-tree.lua [hunk 5]
```diff
@@ -476,6 +502,11 @@ local function validate_options(conf)
 end
 
 function M.setup(conf)
+  if vim.fn.has "nvim-0.7" == 0 then
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:505]: 0.6.1 exits early
0.7.0 works normally

File: lua/nvim-tree/diagnostics.lua [hunk 6]
```diff
@@ -170,8 +170,13 @@ function M.setup(opts)
...
     log.line("diagnostics", "setup")
-    vim.cmd "au DiagnosticChanged * lua require'nvim-tree.diagnostics'.update()"
-    vim.cmd "au User CocDiagnosticChange lua require'nvim-tree.diagnostics'.update()"
+    a.nvim_create_autocmd("DiagnosticChanged", {
+      callback = M.update,
+    })
+    a.nvim_create_autocmd("User", {
+      pattern = "CocDiagnosticChanged",
+      callback = M.update,
+    })
```
Comment (alex-courtis) [inline lua/nvim-tree/diagnostics.lua:179]: DiagnosticChanged works, CocDiagnosticChanged does not.

See https://github.com/kyazdani42/nvim-tree.lua/pull/1218#issuecomment-1114469577 for a test harness you can use.

File: lua/nvim-tree.lua [hunk 7]
```diff
@@ -504,6 +504,7 @@ end
 function M.setup(conf)
   if vim.fn.has "nvim-0.7" == 0 then
     utils.warn "nvim-tree.lua requires Neovim 0.7 or higher"
+    return
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:507]: This is good.

**Review by alex-courtis — CHANGES_REQUESTED** [review]

File: lua/nvim-tree/diagnostics.lua (hunk 6 above)
Comment (sarmong) [reply lua/nvim-tree/diagnostics.lua:179]: @alex-courtis I fix a typo, should work now

**Review by alex-courtis — APPROVED** [review]

**Review by kyazdani42 — APPROVED** [review]

**Pull request merged**